TechnicalDocuments/Proposals/Feedback Navigation.md

```diff
@@ -0,0 +1,81 @@
+# Feedback Navigation
+
+* Author(s): Sergey Shulga
+* Review Manager: TBD
+
+## Introduction
+
+As all of you probably know in our codebase we use `Route` based navigation, where VM has a signal of routes which emits a specific route and FlowController just listens to it and performs navigation.
+
+Although the current concept is fine and we've been happy with it for a while now, it does look a bit weird when it comes to integration with RAF.
+For example, we've recently agreed that we derive all routes from the `State`  of the VM. Also, sometimes we need to pass closures to the routes as we need to notify our VM when a user selects something on the child screen.
+
+```swift
+routes = state.signal.skipRepeats().filterMap { state in
+            switch state {
+            case let .loaded(_, error?):
+                return .showAlert(.make(error: error))
+            case .searching:
+                return .showMap(mapResponsesObserver.send(value:)) // here we are passing a closure which will push value into our VM
+            case let .showingDirections(_, place):
+                return .showDirections(place)
+            }
+        }
+```
+
+## Motivation
+
+The idea is to deal with navigation in the same way we deal with networking. Where navigation can be treated as an async operation the result of which is going to impact the current state of the VM.
+
+## Proposed solution
+
+```swift    
+static func whenSearching(flowController: PharmaciesFlowController) -> Feedback<State, Event> {
+        return Feedback { state -> SignalProducer<Event, NoError> in
+            guard case .searching = state else {
+                return .empty
+            }
+            return flowController.showMap()
+                .map(Event.didSelectPlace)
+        }
+    }
+
+final class PharmaciesFlowController {
+    func showMap() -> SignalProducer<MapResponse, NoError> {}
+    
+    // I case where navigation does not impact current screen
+    func showSomething() -> SignalProducer<Never, NoError> {}
+}
+```
+
+Unlike other Feedbacks, we may not want the navigation to be canceled when the state changes so we can not use `Feedback` overloads which are using `flatMapLatest` under the hood. The possible solution would be to use `flatMapFirst` strategy instead, so if navigation effect is in progress it won't be canceled until it's finished and a new one wont be started.
```

Comment:
> new one wont be started

I think this deserves to mention what will happen if if there is some work in parallel to navigation effect that started earlier and finishes while navigation effect in progress, will it be simply ignored or postponed until navigation effect completes? what if we need behaviour that is different to what `flatMapFirst` does (i.e. if it cancels all parallel work results, is there a way to avoid that if it's needed)

Reply:
> if it cancels all parallel work results, is there a way to avoid that if it's needed

`Feedback`s can not influence other feedbacks. If one uses `flatMapLatest` like most of them than state changes would cancel work in them